### ISSUE TITLE
multi_scan parallel test workers crash when trying to replace Scan.devices

### ISSUE THREAD

**Issue by CyberOwl** [lyndsey-ferguson/fastlane-plugin-test_center#303]
### New Issue Checklist
- [x] Updated `fastlane-plugin-test_center` to the latest version
- [x] I read the [README.md](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/blob/master/README.md)
- [x] I reviewed the [example(s)](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/blob/master/README.md) for the action(s) I am using
- [x] I have removed any sensitive data such as passwords, authentication tokens, or anything else I do not want to world to see

> If you love this fastlane plugin, consider sponsoring it or asking your company to sponsor it. I would really appreciate any
> gesture: https://github.com/sponsors/lyndsey-ferguson. 😍

##### Complete output when running fastlane, including the stack trace and command used

<details>
  <pre>### Captured Output

Command Used: `run_tests --verbose --capture_output`
<details><summary>Output/Log</summary>

```

DEBUG [2020-10-22 10:10:33.53]: Checking if there are any plugins that should be loaded...
DEBUG [2020-10-22 10:10:33.53]: Loading 'fastlane-plugin-test_center' plugin
INFO [2020-10-22 10:10:33.53]: gem 'fastlane-plugin-test_center' is already installed


```

</details>

<details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                                          |
| --------------------------- | -------------------------------------------------------------- |
| OS                          | 10.15.7                                                        |
| Ruby                        | 2.6.3                                                          |
| Bundler?                    | true                                                           |
| Git                         | git version 2.29.0                                             |
| Installation Source         | ~/.gem/bin/fastlane                                            |
| Host                        | Mac OS X 10.15.7 (19H2)                                        |
| Ruby Lib Dir                | /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib |
| OpenSSL Version             | LibreSSL 2.8.3                                                 |
| Is contained                | false                                                          |
| Is homebrew                 | false                                                          |
| Is installed via Fabric.app | false                                                          |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/                    |
| Xcode Version               | 12.1                                                           |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | ru_RU.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby

desc 'Start UI Test session'
lane :run_tests do |options|
  multi_scan(
    project: %any%,
    scheme: %any%,
    testplan: %any%,
    skip_detect_devices: true,
    prelaunch_simulator: false,
    destination: 'platform=iOS Simulator,id=' + @simulator_id,
    output_types: 'xcresult, junit',
    output_files: 'result.xcresult, result.xml',
    retry_test_runner_failures: true,
    result_bundle: true,
    fail_build: true,
    try_count: 3,
    parallel_testrun_count: 4
  )
end

```
</details>

**No Appfile found**

### fastlane gems

| Gem      | Version | Update-Status |
| -------- | ------- | ------------- |
| fastlane | 2.164.0 | ✅ Up-To-Date  |

### Loaded fastlane plugins:

| Plugin                      | Version | Update-Status |
| --------------------------- | ------- | ------------- |
| fastlane-plugin-test_center | 3.14.6  | ✅ Up-To-Date  |

<details><summary><b>Loaded gems</b></summary>

| Gem                         | Version      |
| --------------------------- | ------------ |
| did_you_mean                | 1.4.0        |
| bundler                     | 2.1.4        |
| rake                        | 13.0.1       |
| CFPropertyList              | 3.0.2        |
| public_suffix               | 4.0.6        |
| addressable                 | 2.7.0        |
| atomos                      | 0.1.3        |
| aws-eventstream             | 1.1.0        |
| aws-partitions              | 1.384.0      |
| aws-sigv4                   | 1.2.2        |
| jmespath                    | 1.4.0        |
| aws-sdk-core                | 3.109.1      |
| aws-sdk-kms                 | 1.39.0       |
| aws-sdk-s3                  | 1.83.1       |
| babosa                      | 1.0.4        |
| claide                      | 1.0.3        |
| colored                     | 1.2          |
| colored2                    | 3.1.2        |
| colorize                    | 0.8.1        |
| highline                    | 1.7.10       |
| commander-fastlane          | 4.4.6        |
| declarative                 | 0.0.20       |
| declarative-option          | 0.1.0        |
| digest-crc                  | 0.6.1        |
| unf_ext                     | 0.0.7.7      |
| unf                         | 0.1.4        |
| domain_name                 | 0.5.20190701 |
| dotenv                      | 2.7.6        |
| emoji_regex                 | 3.0.0        |
| excon                       | 0.78.0       |
| multipart-post              | 2.0.0        |
| ruby2_keywords              | 0.0.2        |
| faraday                     | 1.1.0        |
| http-cookie                 | 1.0.3        |
| faraday-cookie_jar          | 0.0.7        |
| faraday_middleware          | 1.0.0        |
| fastimage                   | 2.2.0        |
| gh_inspector                | 1.1.3        |
| jwt                         | 2.2.2        |
| memoist                     | 0.16.2       |
| multi_json                  | 1.15.0       |
| os                          | 1.1.1        |
| signet                      | 0.14.0       |
| googleauth                  | 0.14.0       |
| httpclient                  | 2.8.3        |
| mini_mime                   | 1.0.2        |
| uber                        | 0.1.0        |
| representable               | 3.0.4        |
| retriable                   | 3.1.2        |
| google-api-client           | 0.38.0       |
| google-cloud-env            | 1.4.0        |
| google-cloud-errors         | 1.0.1        |
| google-cloud-core           | 1.5.0        |
| google-cloud-storage        | 1.29.1       |
| json                        | 2.3.1        |
| mini_magick                 | 4.10.1       |
| plist                       | 3.5.0        |
| rubyzip                     | 2.3.0        |
| security                    | 0.1.3        |
| naturally                   | 2.2.0        |
| simctl                      | 1.6.8        |
| slack-notifier              | 2.3.2        |
| terminal-notifier           | 2.0.0        |
| unicode-display_width       | 1.7.0        |
| terminal-table              | 1.8.0        |
| tty-screen                  | 0.8.1        |
| tty-cursor                  | 0.7.1        |
| tty-spinner                 | 0.9.3        |
| word_wrap                   | 1.0.0        |
| nanaimo                     | 0.3.0        |
| xcodeproj                   | 1.19.0       |
| rouge                       | 2.0.7        |
| xcpretty                    | 0.3.0        |
| xcpretty-travis-formatter   | 1.0.0        |
| trainer                     | 0.9.1        |
| xctest_list                 | 1.2.1        |
| fastlane-plugin-test_center | 3.14.6       |
</details>

*generated on:* **2020-10-22**
</details></pre>
</details>

**Issue:**
- Cannot reproduce with `fastlane ver. "2.163.0"` & `fastlane-plugin-test_center ver. "3.14.4"`
- Works fine when setting `collate_reports` option to `false` and removing `xcresult` from `output_types`

**Failure log:**
<details><pre>
[worker 4] undefined method `replace' for nil:NilClass
[worker 4] ["/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/retrying_scan.rb:53:in `update_scan_options'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/retrying_scan.rb:67:in `run'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/retrying_scan.rb:59:in `run'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/test_batch_worker.rb:18:in `run'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/parallel_test_batch_worker.rb:49:in `block in run'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/parallel_test_batch_worker.rb:45:in `fork'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/parallel_test_batch_worker.rb:45:in `run'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:239:in `block in run_test_batches'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:236:in `each'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:236:in `each_with_index'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:236:in `run_test_batches'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:107:in `block in run'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:134:in `block in setup_run_tests_for_each_device'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:123:in `each'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:123:in `each_with_index'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:123:in `setup_run_tests_for_each_device'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:105:in `run'", "/Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/actions/multi_scan.rb:43:in `run'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:263:in `block (2 levels) in execute_action'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/actions/actions_helper.rb:69:in `execute_action'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:255:in `block in execute_action'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:229:in `chdir'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:229:in `execute_action'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:157:in `trigger_action_by_name'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'", "Fastfile:90:in `block in parsing_binding'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/lane.rb:33:in `call'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:204:in `try_switch_to_lane'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:146:in `trigger_action_by_name'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'", "Fastfile:53:in `block in parsing_binding'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/lane.rb:33:in `call'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:49:in `block in execute'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:45:in `chdir'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:45:in `execute'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/lane_manager.rb:47:in `cruise_lane'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/command_line_handler.rb:36:in `handle'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/commands_generator.rb:108:in `block (2 levels) in run'", "/Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'", "/Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'", "/Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'", "/Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/commands_generator.rb:352:in `run'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/commands_generator.rb:41:in `start'", "/Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/cli_tools_distributor.rb:119:in `take_off'", "/Users/username/.gem/gems/fastlane-2.164.0/bin/fastlane:23:in `<top (required)>'", "/Users/username/.gem/bin/fastlane:23:in `load'", "/Users/username/.gem/bin/fastlane:23:in `<top (required)>'", "/Library/Ruby/Gems/2.6.0/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:63:in `load'", "/Library/Ruby/Gems/2.6.0/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:63:in `kernel_load'", "/Library/Ruby/Gems/2.6.0/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:28:in `run'", "/Library/Ruby/Gems/2.6.0/gems/bundler-2.1.4/lib/bundler/cli.rb:476:in `exec'", "/Library/Ruby/Gems/2.6.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'", "/Library/Ruby/Gems/2.6.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'", "/Library/Ruby/Gems/2.6.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor.rb:399:in `dispatch'", "/Library/Ruby/Gems/2.6.0/gems/bundler-2.1.4/lib/bundler/cli.rb:30:in `dispatch'", "/Library/Ruby/Gems/2.6.0/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/base.rb:476:in `start'", "/Library/Ruby/Gems/2.6.0/gems/bundler-2.1.4/lib/bundler/cli.rb:24:in `start'", "/Library/Ruby/Gems/2.6.0/gems/bundler-2.1.4/exe/bundle:46:in `block in <top (required)>'", "/Library/Ruby/Gems/2.6.0/gems/bundler-2.1.4/lib/bundler/friendly_errors.rb:123:in `with_friendly_errors'", "/Library/Ruby/Gems/2.6.0/gems/bundler-2.1.4/exe/bundle:34:in `<top (required)>'", "/Users/username/.gem/bin/bundle:23:in `load'", "/Users/username/.gem/bin/bundle:23:in `<main>'"]
[worker 4] DEBUG [2020-10-22 10:04:05.20]: All tests passed for batch 4? false
DEBUG [2020-10-22 10:04:33.56]: Collating results for all batches
DEBUG [2020-10-22 10:04:33.56]: ReportCollator collating
INFO [2020-10-22 10:04:33.56]: ------------------------------------------------------------------------
INFO [2020-10-22 10:04:33.56]: --- Step: xcrun simctl delete 'C04F5ABA-BDC1-44FE-8641-6CB953FC72C5' ---
INFO [2020-10-22 10:04:33.56]: ------------------------------------------------------------------------
INFO [2020-10-22 10:04:33.56]: $ xcrun simctl delete 'C04F5ABA-BDC1-44FE-8641-6CB953FC72C5'
INFO [2020-10-22 10:04:33.74]: ▸ Using Parallel Testing Device Clones Device Set: '/Users/username/Library/Developer/XCTestDevices'
WARN [2020-10-22 10:04:33.86]: Lane Context:
INFO [2020-10-22 10:04:33.86]: {:PLATFORM_NAME=>nil, :LANE_NAME=>"run_tests"}
ERROR [2020-10-22 10:04:33.86]: No such file or directory @ rb_file_s_symlink - (/Users/username/project/UITests/fastlane/test_output/UITests/result.xcresult, /Users/username/project/UITests/fastlane/test_output/UITests/result.test_result)
INFO [2020-10-22 10:04:33.86]: Successfully generated documentation at path '/Users/username/project/UITests/fastlane/README.md'

+------+---------------------------------------------------------------+-------------+
|                                  fastlane summary                                  |
+------+---------------------------------------------------------------+-------------+
| Step | Action                                                        | Time (in s) |
+------+---------------------------------------------------------------+-------------+
| 1    | Switch to test_run lane                                       | 0           |
| 💥   | multi_scan                                                    | 112         |
| 3    | xcrun simctl delete 'C04F5ABA-BDC1-44FE-8641-6CB953FC72C5'    | 0           |
+------+---------------------------------------------------------------+-------------+

DEBUG [2020-10-22 10:04:33.91]: All plugins are up to date
ERROR [2020-10-22 10:04:33.91]: fastlane finished with errors

WARN [2020-10-22 10:04:33.91]: Error accessing file, this might be due to fastlane's directory handling
WARN [2020-10-22 10:04:33.91]: Check out https://docs.fastlane.tools/advanced/#directory-behavior for more details

DEBUG [2020-10-22 10:04:33.91]: All plugins are up to date
bundler: failed to load command: fastlane (/Users/username/.gem/bin/fastlane)
Errno::ENOENT: No such file or directory @ rb_file_s_symlink - (/Users/username/project/UITests/fastlane/test_output/UITests/result.xcresult, /Users/username/project/UITests/fastlane/test_output/UITests/result.test_result)
  /Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:332:in `symlink'
  /Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:332:in `symlink_result_bundle_to_xcresult'
  /Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:390:in `collate_batched_reports_for_testable'
  /Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:271:in `block in collate_batched_reports'
  /Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:270:in `each'
  /Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:270:in `collate_batched_reports'
  /Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:242:in `run_test_batches'
  /Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:107:in `block in run'
  /Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:134:in `block in setup_run_tests_for_each_device'
  /Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:123:in `each'
  /Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:123:in `each_with_index'
  /Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:123:in `setup_run_tests_for_each_device'
  /Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:105:in `run'
  /Users/username/.gem/gems/fastlane-plugin-test_center-3.14.6/lib/fastlane/plugin/test_center/actions/multi_scan.rb:43:in `run'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:263:in `block (2 levels) in execute_action'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/actions/actions_helper.rb:69:in `execute_action'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:255:in `block in execute_action'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:229:in `chdir'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:229:in `execute_action'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:157:in `trigger_action_by_name'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'
  Fastfile:90:in `block in parsing_binding'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/lane.rb:33:in `call'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:204:in `try_switch_to_lane'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:146:in `trigger_action_by_name'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'
  Fastfile:53:in `block in parsing_binding'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/lane.rb:33:in `call'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:49:in `block in execute'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:45:in `chdir'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:45:in `execute'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/lane_manager.rb:47:in `cruise_lane'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/command_line_handler.rb:36:in `handle'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/commands_generator.rb:108:in `block (2 levels) in run'
  /Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
  /Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
  /Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
  /Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/commands_generator.rb:352:in `run'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/commands_generator.rb:41:in `start'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/cli_tools_distributor.rb:119:in `take_off'
  /Users/username/.gem/gems/fastlane-2.164.0/bin/fastlane:23:in `<top (required)>'
  /Users/username/.gem/bin/fastlane:23:in `load'
  /Users/username/.gem/bin/fastlane:23:in `<top (required)>'
INFO [2020-10-22 10:04:33.91]: Deleting UI-Tests Clone 1 TestCenter::Helper::MultiScanManager::SimulatorHelper<70110723876480>
INFO [2020-10-22 10:04:34.08]: Deleting UI-Tests Clone 2 TestCenter::Helper::MultiScanManager::SimulatorHelper<70110723876480>
INFO [2020-10-22 10:04:34.22]: Deleting UI-Tests Clone 3 TestCenter::Helper::MultiScanManager::SimulatorHelper<70110723876480>
INFO [2020-10-22 10:04:34.36]: Deleting UI-Tests Clone 4 TestCenter::Helper::MultiScanManager::SimulatorHelper<70110723876480>
</pre></details>

**Comment by lyndsey-ferguson** [comment]
@CyberOwl, Can you modify your `Pluginfile` per my instructions below, run `bundle install`, and then run your fastlane again (with the `--verbose` flag)?

Pluginfile:
```
gem 'fastlane-plugin-test_center', :git => "https://github.com/lyndsey-ferguson/fastlane-plugin-test_center.git", :branch => "issue-303-no-scan-devices-crash"
```

If there are still problems, please let me know and attach the console output as a text file to this issue (makes it easier for me to review).  If it works, please let me know.

**Comment by CyberOwl** [comment]
With branch `issue-303-no-scan-devices-crash` multi_scan starts idling (w/o running tests) after ` Starting test run N` message

<details><pre>
INFO [2020-10-22 16:22:55.70]: Testing batches for device 'UI-Tests'
DEBUG [2020-10-22 16:22:55.71]: Fetching available simulator devices
INFO [2020-10-22 16:22:56.08]: Deleting UI-Tests Clone 1 TestCenter::Helper::MultiScanManager::SimulatorHelper<70247765890480>
INFO [2020-10-22 16:22:56.26]: Deleting UI-Tests Clone 2 TestCenter::Helper::MultiScanManager::SimulatorHelper<70247765890480>
INFO [2020-10-22 16:22:56.26]: Shutting down D0779E96-9C5C-484B-B9B9-001261A18144
DEBUG [2020-10-22 16:23:02.68]: Fetching available simulator devices
INFO [2020-10-22 16:23:02.98]: Shutting down 40769C03-E7A3-4CD6-906B-6F1F4C882508
DEBUG [2020-10-22 16:23:12.17]: Cloned simulator UI-Tests to (name=UI-Tests Clone 1, udid=87E74D10-9167-42B3-BDFC-F563272E0C43, OS=14.1)
DEBUG [2020-10-22 16:23:15.52]: Cloned simulator UI-Tests to (name=UI-Tests Clone 2, udid=252BD7A7-4A33-49AB-B2EE-E1A9F9BEB7E9, OS=14.1)
INFO [2020-10-22 16:23:15.69]: Booting UI-Tests Clone 1 TestCenter::Helper::MultiScanManager::SimulatorHelper<70278337645080>
INFO [2020-10-22 16:23:16.26]: Booting UI-Tests Clone 2 TestCenter::Helper::MultiScanManager::SimulatorHelper<70278337645080>
INFO [2020-10-22 16:23:17.30]: Starting test run 1
INFO [2020-10-22 16:23:17.30]: Starting test run 2


**Interrupt signal**

INFO [2020-10-22 16:23:45.50]: ------------------------------------------------------------------------
INFO [2020-10-22 16:23:45.50]: --- Step: xcrun simctl delete '40769C03-E7A3-4CD6-906B-6F1F4C882508' ---
INFO [2020-10-22 16:23:45.50]: ------------------------------------------------------------------------
INFO [2020-10-22 16:23:45.50]: xcrun simctl delete '40769C03-E7A3-4CD6-906B-6F1F4C882508'
INFO [2020-10-22 16:23:45.79]: ---------------------------------------------------
INFO [2020-10-22 16:23:45.79]: --- Step: xcrun simctl --set testing delete all ---
INFO [2020-10-22 16:23:45.79]: ---------------------------------------------------
INFO [2020-10-22 16:23:45.79]: xcrun simctl --set testing delete all
INFO [2020-10-22 16:23:45.84]: ▸ [0;35;49mUsing Parallel Testing Device Clones Device Set: '/Users/username/Library/Developer/XCTestDevices'
WARN [2020-10-22 16:23:45.96]: [0;33;49mLane Context:
INFO [2020-10-22 16:23:45.96]: {:PLATFORM_NAME=>nil, :LANE_NAME=>"run_ui_tests_client"}
ERROR [2020-10-22 16:23:45.96]: Error in your Fastfile at line 154
ERROR [2020-10-22 16:23:45.96]:     152:	# -----------------------------------------------------------------
ERROR [2020-10-22 16:23:45.96]:     153:	
ERROR [2020-10-22 16:23:45.96]:  => 154:	Signal.trap('INT') { throw :force_quit }
ERROR [2020-10-22 16:23:45.96]:     155:	Signal.trap('TERM') { throw :force_quit }
ERROR [2020-10-22 16:23:45.96]:     156:	
ERROR [2020-10-22 16:23:45.96]: uncaught throw :force_quit
INFO [2020-10-22 16:23:45.96]: Successfully generated documentation at path '/Users/username/target/UITests/fastlane/README.md'
DEBUG [2020-10-22 16:23:45.98]: All plugins are up to date
ERROR [2020-10-22 16:23:45.98]: fastlane finished with errors

Looking for related GitHub issues on fastlane/fastlane...
Search query: uncaught throw :force_quit

URL: https://api.github.com/search/issues?q=uncaught%20throw%20%3Aforce_quit+repo:fastlane/fastlane
Found no similar issues. To create a new issue, please visit:
https://github.com/fastlane/fastlane/issues/new
Run `fastlane env` to append the fastlane environment to your issue
DEBUG [2020-10-22 16:23:46.44]: All plugins are up to date
INFO [2020-10-22 16:23:46.45]: Deleting UI-Tests Clone 1 TestCenter::Helper::MultiScanManager::SimulatorHelper<70278337645080>
INFO [2020-10-22 16:23:46.45]: Shutting down 87E74D10-9167-42B3-BDFC-F563272E0C43
INFO [2020-10-22 16:23:52.86]: Deleting UI-Tests Clone 2 TestCenter::Helper::MultiScanManager::SimulatorHelper<70278337645080>
INFO [2020-10-22 16:23:52.86]: Shutting down 252BD7A7-4A33-49AB-B2EE-E1A9F9BEB7E9

**Interrupt signal**

bundler: failed to load command: fastlane (/Users/username/.gem/bin/fastlane)
UncaughtThrowError: [!] uncaught throw :force_quit
  Fastfile:154:in `throw'
  Fastfile:154:in `block in parsing_binding'
  /Users/username/.gem/bundler/gems/fastlane-plugin-test_center-a8a6f5574f14/lib/fastlane/plugin/test_center/helper/multi_scan_manager/test_batch_worker_pool.rb:134:in `wait'
  /Users/username/.gem/bundler/gems/fastlane-plugin-test_center-a8a6f5574f14/lib/fastlane/plugin/test_center/helper/multi_scan_manager/test_batch_worker_pool.rb:134:in `block in wait_for_all_workers'
  /Users/username/.gem/bundler/gems/fastlane-plugin-test_center-a8a6f5574f14/lib/fastlane/plugin/test_center/helper/multi_scan_manager/test_batch_worker_pool.rb:133:in `each'
  /Users/username/.gem/bundler/gems/fastlane-plugin-test_center-a8a6f5574f14/lib/fastlane/plugin/test_center/helper/multi_scan_manager/test_batch_worker_pool.rb:133:in `wait_for_all_workers'
  /Users/username/.gem/bundler/gems/fastlane-plugin-test_center-a8a6f5574f14/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:241:in `run_test_batches'
  /Users/username/.gem/bundler/gems/fastlane-plugin-test_center-a8a6f5574f14/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:107:in `block in run'
  /Users/username/.gem/bundler/gems/fastlane-plugin-test_center-a8a6f5574f14/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:134:in `block in setup_run_tests_for_each_device'
  /Users/username/.gem/bundler/gems/fastlane-plugin-test_center-a8a6f5574f14/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:123:in `each'
  /Users/username/.gem/bundler/gems/fastlane-plugin-test_center-a8a6f5574f14/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:123:in `each_with_index'
  /Users/username/.gem/bundler/gems/fastlane-plugin-test_center-a8a6f5574f14/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:123:in `setup_run_tests_for_each_device'
  /Users/username/.gem/bundler/gems/fastlane-plugin-test_center-a8a6f5574f14/lib/fastlane/plugin/test_center/helper/multi_scan_manager/runner.rb:105:in `run'
  /Users/username/.gem/bundler/gems/fastlane-plugin-test_center-a8a6f5574f14/lib/fastlane/plugin/test_center/actions/multi_scan.rb:43:in `run'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:263:in `block (2 levels) in execute_action'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/actions/actions_helper.rb:69:in `execute_action'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:255:in `block in execute_action'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:229:in `chdir'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:229:in `execute_action'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:157:in `trigger_action_by_name'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'
  Fastfile:90:in `block in parsing_binding'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/lane.rb:33:in `call'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:204:in `try_switch_to_lane'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:146:in `trigger_action_by_name'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/fast_file.rb:159:in `method_missing'
  Fastfile:53:in `block in parsing_binding'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/lane.rb:33:in `call'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:49:in `block in execute'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:45:in `chdir'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/runner.rb:45:in `execute'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/lane_manager.rb:47:in `cruise_lane'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/command_line_handler.rb:36:in `handle'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/commands_generator.rb:108:in `block (2 levels) in run'
  /Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
  /Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
  /Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
  /Library/Ruby/Gems/2.6.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/commands_generator.rb:352:in `run'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/commands_generator.rb:41:in `start'
  /Users/username/.gem/gems/fastlane-2.164.0/fastlane/lib/fastlane/cli_tools_distributor.rb:119:in `take_off'
  /Users/username/.gem/gems/fastlane-2.164.0/bin/fastlane:23:in `<top (required)>'
  /Users/username/.gem/bin/fastlane:23:in `load'
  /Users/username/.gem/bin/fastlane:23:in `<top (required)>'
</pre></details>

**Comment by lyndsey-ferguson** [comment]
It looks like a problem in your Fastfile:
```
UncaughtThrowError: [!] uncaught throw :force_quit

Fastfile:154:in throw' Fastfile:154:in block in parsing_binding'
```

**Comment by CyberOwl** [comment]
No, i just manually send `SIGINT` to terminate process and handle it with `Signal.trap('INT') { throw :force_quit }`

**Comment by lyndsey-ferguson** [comment]
```
INFO [2020-10-22 16:23:17.30]: Starting test run 1
INFO [2020-10-22 16:23:17.30]: Starting test run 2
```
means that the tests are running in parallel.  How long do you let it run for?

**Comment by CyberOwl** [comment]
After 3~ minutes tests has been started. After 1 minute testrun just freezes for long time and doesn't respond.

Comparison: fastlane & test_center from branch: 30 min 
Latest fastlane & test_center:

**Comment by CyberOwl** [comment]
- Idle after simulator initialization like 4~ minutes
- Idle after some tests (XCTest TestRunner doesn't respond too)
- Test run completes after 40~ minutes with 4 parallel runs and 16 tests in total

**Comment by lyndsey-ferguson** [comment]
Can you summarize what you're trying to tell me? 

> upd. Running tests with plugin causes OS stuttering and performance issues :(

Keep in mind that using more simulators will put more load on your computer. Try with a parallel_testrun_count of 2.

**Comment by CyberOwl** [comment]
I've tested pure Xcode 12 parallel testing and got same performance issues which i never I have not seen before.

**Comment by CyberOwl** [comment]
So... initial issue (`issue-303-no-scan-devices-crash`) is fixed. Thank you!

**Comment by lyndsey-ferguson** [comment]
> I've tested pure Xcode 12 parallel testing and got same performance issues which i never I have not seen before.

So, you're saying that you're seeing the same stuttering and performance issues? What sort of specs do you have? I have been able to go up to 8 parallel test runs without much of an issue...(on a MacBook Pro with 16GB of RAM).

You may also want to make sure that you don't have left over processes that are taking up power...reboot the comptuer?

**Comment by lyndsey-ferguson** [comment]
Fixed in [v3.14.7](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center/releases/tag/v3.14.7)
